tests/test_tensor/test_zero_optim.py

```diff
@@ -79,44 +75,53 @@ def run_gpt(use_chunk, use_zero, placement_policy, tp_init_spec_func=None):
     for torch_p, p in zip(torch_model.parameters(), model.parameters()):
         torch_p.data.copy_(p)
 
+    world_size = torch.distributed.get_world_size()
+
+    # world size, dp = 2, tp =2, construct a hybrid parallelism.
+    if world_size == 4:
+        pg = ProcessGroup(tp_degree=2)
+    else:
+        pg = ProcessGroup(tp_degree=world_size)
+
     if tp_init_spec_func:
-        tp_init_spec_func(model)
+        tp_init_spec_func(model, pg)
 
     chunk_size = ChunkManager.search_chunk_size(model, 8192, 8) if use_chunk else None
     chunk_manager = ChunkManager(chunk_size,
                                  enable_distributed_storage=use_zero,
                                  init_device=GeminiManager.get_default_device(placement_policy))
     gemini_manager = GeminiManager(placement_policy, chunk_manager)
-    model = ZeroDDP(model, gemini_manager)
+    model = ZeroDDP(model, gemini_manager, pg)
     optim = HybridAdam(model.parameters(), lr=1e-3)
     optim = ZeroOptimizer(optim, model, initial_scale=32)
 
     amp_config = dict(opt_level='O2', keep_batchnorm_fp32=False, loss_scale=32)
     torch_optim = torch.optim.Adam(torch_model.parameters(), lr=1e-3)
     torch_model, torch_optim = convert_to_apex_amp(torch_model, torch_optim, amp_config)
-    torch_model = DDP(torch_model, device_ids=[gpc.get_global_rank()], process_group=gpc.get_group(ParallelMode.DATA))
+    torch_model = DDP(torch_model, device_ids=[pg.rank()], process_group=pg.dp_process_group())
 
-    print(chunk_manager)
-    check_param_equal(model, torch_model)
+    # print(chunk_manager)
+    check_param_equal(model, torch_model, pg)
     model.train()
     torch_model.train()
-    set_seed(gpc.get_local_rank(ParallelMode.DATA))
+    set_seed(pg.dp_local_rank())
     for i, (input_ids, attn_mask) in enumerate(train_dataloader):
         if i > 2:
             break
+
+        torch.distributed.broadcast(input_ids, 0)
```

Comment:
Don't broadcast. We except DP ranks have different inputs.

Reply:
OK